examples/tpu_training-tf.ipynb

```diff
@@ -0,0 +1,904 @@
+{
```

Comment:
<p>Insist on the fact that you need a fixed length for TPUs, which is why you pad to max length.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/notebooks/pull/313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Reply:
<p>I think variable lengths are possible, at the cost of a new XLA compilation each time! (This is definitely the case with standard XLA in TF, but I haven't benchmarked it extensively with TPUs)</p><p> </p>

Reply:
Yes, it recompiles each time, which makes it unusable in practice ;-)

Reply:
<p>In our other TF-XLA notebooks we recommend using <code>pad_to_multiple_of</code> with a multiple like 32 or 64 - this means there's only a few unique shapes, but you still get a big performance boost compared to max length padding or running without XLA. I'll check if that works on TPU too!</p><p> </p>

Reply:
> In our other TF-XLA notebooks we recommend using pad_to_multiple_of with a multiple like 32 or 64 - this means there's only a few unique shapes, but you still get a big performance boost compared to max length padding or running without XLA. I'll check if that works on TPU too!

You mean checking if shorter sequence lengths like 32 or 64 help boost the speed? @Rocketknight1

Reply:
<p>Yes, but also that XLA compilation and cacheing works the same way. </p><p> </p><p>When running TF normally, multiple XLA compilations can be cached. For example, if you pad to a multiple of 64 and so your dataset only contains three unique batch shapes, for example (8, 128), (8, 192) and (8, 256), then TF will run an XLA compilation for each of those shapes the first time it encounters them, but after that it will use the cached XLA programs and training will be fast.</p><p> </p><p>I'm going to check if the same thing works on TPU!</p>

Reply:
<p>Just tested it and confirmed it works the same way!</p>

Reply:
<p>I'll explain this in more detail in the XLA section of the tutorial doc, which will be linked once it's uploaded.</p>